### PR TITLE
Changes for 1.2.2

### PR DESCRIPTION
**Pull request by jvinolas**
### Added

- Hypervisor autodetects VLANS or can be set by script.
- Adding remote hypervisor through CLI has all the UI options now.
  docker exec -e ENABLED=True -e ID=isard-hypervisor -e HYPERVISOR=192.168.0.10 -e PASSWORD=isard -e PORT=2022 -e POOL=default -e DISKOP=True -e VIRTUALOP=True -e VIEWERHOST=localhost -e VIEWERNATHOST=isardvdi.com -e VIEWERNATOFFSET=30000 isard-app sh -c '/add-hypervisor.sh'
- Added squid image that allows proxying spice viewers.
- Now .env file allows to configure main system parameters.
- Grafana container updated to latest version and uses .env file
- Rsync added to Hypervisor image so now can handle different mount points.
- Maintenance mode (on/off) in nginx. docker exec -ti isard-nginx /bin/sh -c "maintenance.sh on"

### Fixed

- Fixed supervisord parameters in app to be more stable at restart.
- Hypervisor default network is forced to be active as failed in some systems.
- Mosquitto extras image rebuild and now automatically monitors Espurna IoT power devices.
- Removed rsync bwlimit when creating templates

### Changed
- Grafana is now part of main docker-compose file.
- Database upgrade modified python3 rethinkdb connection

### Removed
-

### CODE REVIEW

File: dockers/app/Dockerfile [hunk 1]
```diff
@@ -1,34 +1,41 @@
-FROM alpine:3.8 as production
+FROM alpine:3.11.3 as test
```
Comment (NefixEstrada) [inline dockers/app/Dockerfile]: Why is the stage named `test`?

File: dockers/hypervisor/Dockerfile [hunk 2]
```diff
@@ -17,7 +17,8 @@ RUN apk add --no-cache \
 RUN apk add --no-cache --virtual .build_deps \
     build-base \
     python3-dev 
-RUN pip3 install --no-cache-dir websockify==0.8.0
+RUN pip3 install --no-cache-dir websockify
```
Comment (NefixEstrada) [inline dockers/hypervisor/Dockerfile]: This should be `websockify=0.9.0`, because in a future it might break

File: dockers/hypervisor/start_proxy.py [hunk 3]
```diff
@@ -3,14 +3,20 @@
 import multiprocessing
 import websockify
 import socket
+import os
+
+try:
+    websockets = int(os.getenv('WEBSOCKETS','10'))
+except:
+    websockets=10
```
Comment (NefixEstrada) [inline dockers/hypervisor/start_proxy.py]: Why changing the default value to 10 instead of 50?

File: dockers/nginx/Dockerfile [hunk 4]
```diff
@@ -10,3 +10,5 @@ COPY dockers/nginx/dh.pem /
 COPY dockers/nginx/auto-generate-certs.sh /opt
 COPY dockers/nginx/entrypoint.sh /opt/
 RUN chmod 744 /opt/*
+
+FROM production as development
```
Comment (NefixEstrada) [inline dockers/nginx/Dockerfile]: Why changing the name of the stage to `development`?

File: .env.original [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+##       then be routed through proxy port.
+SPICE_HOSTS=false
+SPICE_PROXY_PORT=5899
+
+# HTML5 WEBSOCKETS
+## All HTML5 traffic is done through 443 port and here you can
+## set how many websockets you will need for web desktops viewers.
+## Consider 4 ports for each domain
+HTML5_WEBSOCKETS=100
+
```
Comment (NefixEstrada) [inline .env.original]: What is this file?

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,33 @@
...
+- Fixed supervisord parameters in app to be more stable at restart.
+- Hypervisor default network is forced to be active as failed in some systems.
+- Mosquitto extras image rebuild and now automatically monitors Espurna IoT power devices.
+- Removed rsync bwlimit when creating templates
+
+### Changed
+- Grafana is now part of main docker-compose file.
+- Database upgrade modified python3 rethinkdb connection
+
+### Removed
```
Comment (NefixEstrada) [inline CHANGELOG.md]: If nothing was removed, it's not necessary to have it in the changelog

File: docker-compose.yml [hunk 7]
```diff
@@ -30,24 +30,29 @@ services:
...
     ports:
-      - "5900-5999:5900-5999"
-      - "55900-55999:55900-55999"
+      - "2022:22"
+      - "5900-6199:5900-6199"
+      - "6400-6699:6400-6699"
     networks:
       - isard_network
+#      - default
+#      - internet-vms
```
Comment (NefixEstrada) [inline docker-compose.yml]: This comments could be deleted, since they're not used for anything

File: docker-compose.yml [hunk 8]
```diff
@@ -72,7 +79,53 @@ services:
...
   isard_network:
     external: false
     name: isard_network
+#  internet-vms:
+#    driver: macvlan
+#    driver_opts:
+#      parent: viewers
+#    ipam:
+#      config:
+#        - subnet: 172.30.0.0/16
```
Comment (NefixEstrada) [inline docker-compose.yml]: This comments could be deleted, since they're not used for anything

File: dockers/app/Dockerfile [hunk 9]
```diff
@@ -1,34 +1,41 @@
...
+        libxml2-dev \
+        libxslt-dev \
+	gcc
 
+RUN pip3 install --upgrade pip
+RUN pip3 install --no-cache-dir pandas
 COPY dockers/app/requirements.pip3 /requirements.pip3
 RUN pip3 install --no-cache-dir -r requirements.pip3
+#RUN pip3 list --outdated --format=freeze | grep -v '^\-e' | cut -d = -f 1 | xargs -n1 pip3 install -U && exit 0
+#RUN pip3 install rethinkdb==2.3.0.post6  werkzeug==0.16.0 APScheduler==3.3.1 python-engineio==3.8.1
```
Comment (NefixEstrada) [inline dockers/app/Dockerfile]: This comments could be deleted, since they're not used for anything

File: dockers/app/Dockerfile [hunk 10]
```diff
@@ -44,14 +51,21 @@ RUN mv /isard/isard.conf.docker /isard/isard.conf
...
 EXPOSE 5000
 
 CMD ["/usr/bin/supervisord", "-c", "/etc/supervisord.conf"]
 
+
 FROM production as development
 RUN pip3 install --no-cache-dir ipython pytest
-RUN apk add --no-cache --update bash vim openssh bash
+RUN apk add --no-cache --update bash vim openssh 
+#py3-pexpect
```
Comment (NefixEstrada) [inline dockers/app/Dockerfile]: This comments could be deleted, since they're not used for anything

File: dockers/app/Dockerfile [hunk 11]
```diff
@@ -44,14 +51,21 @@ RUN mv /isard/isard.conf.docker /isard/isard.conf
...
 COPY dockers/app/certs.sh /
 COPY dockers/app/add-hypervisor.sh /
+COPY dockers/app/add-hyper-rethink.py /
 COPY dockers/app/supervisord.conf /etc/supervisord.conf
 
 EXPOSE 5000
 
 CMD ["/usr/bin/supervisord", "-c", "/etc/supervisord.conf"]
 
+
```
Comment (NefixEstrada) [inline dockers/app/Dockerfile]: This newline is useless

File: dockers/app/requirements.pip3 [hunk 12]
```diff
@@ -1,9 +1,67 @@
...
+chardet==3.0.4
+Click==7.0
+cryptography==2.8
+decorator==4.4.1
+Flask==0.12.5
+Flask-Login==0.5.0
 Flask-SocketIO==2.8.6
-iniparse==0.4
-rethinkdb==2.3.0.post6
+###Flask-SocketIO==4.2.1
```
Comment (NefixEstrada) [inline dockers/app/requirements.pip3]: This comments could be deleted, since they're not used for anything

File: dockers/app/Dockerfile_quefunciona_antics [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+EXPOSE 5000
+
+CMD ["/usr/bin/supervisord", "-c", "/etc/supervisord.conf"]
+
+FROM production as development
+RUN pip3 install --no-cache-dir ipython pytest
+RUN apk add --no-cache --update bash vim openssh bash
+
+CMD ["/usr/bin/supervisord", "-c", "/etc/supervisord.conf"]
+
```
Comment (NefixEstrada) [inline dockers/app/Dockerfile_quefunciona_antics]: This file is unnecessary, it is stored in the git history

File: dockers/app/Dockerfile_upgrade [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+EXPOSE 5000
+
+CMD ["/usr/bin/supervisord", "-c", "/etc/supervisord.conf"]
+
+
+FROM test as development
+RUN pip3 install --no-cache-dir ipython pytest
+RUN apk add --no-cache --update bash vim openssh bash py3-pexpect
+
+CMD ["/usr/bin/supervisord", "-c", "/etc/supervisord.conf"]
```
Comment (NefixEstrada) [inline dockers/app/Dockerfile_upgrade]: What is this Dockerfile?

File: dockers/app/requirements.pip3.original [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+python-engineio==3.8.1
+python-socketio==4.1.0
+pytz==2019.1
+requests==2.19.1
+rethinkdb==2.3.0.post6
+six==1.11.0
+tzlocal==1.5.1
+urllib3==1.22
+Werkzeug==0.11.15
+xmltodict==0.11.0
```
Comment (NefixEstrada) [inline dockers/app/requirements.pip3.original]: Is this file necessary? It could be fetched through the git history

File: dockers/app/requirements.pip3.quefunciona_antics [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+Flask-SocketIO==2.8.6
+iniparse==0.4
+rethinkdb==2.3.0.post6
+pynpm==0.1.1
+graphyte==1.4
+pem==18.2.0
+Flask-Login==0.4.1
+xmltodict==0.11.0
+python-engineio==3.8.1
+python-socketio==4.1.0
```
Comment (NefixEstrada) [inline dockers/app/requirements.pip3.quefunciona_antics]: Is this file necessary? It could be fetched through git history

File: dockers/app/upgraded-req.pip3 [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+requests==2.23.0
+rethinkdb==2.3.0.post6
+six==1.14.0
+supervisor==4.1.0
+traitlets==4.3.3
+tzlocal==2.0.0
+urllib3==1.25.8
+wcwidth==0.1.8
+Werkzeug==0.16.0
+xmltodict==0.12.0
```
Comment (NefixEstrada) [inline dockers/app/upgraded-req.pip3]: Why is this file for?

File: dockers/grafana/Dockerfile [hunk 18]
```diff
@@ -34,16 +34,16 @@ RUN apk add --no-cache \
   libffi-dev
 
 # Grafana
-ENV GRAFANA_VERSION=5.4.3
-ENV GRAFANA_VERSION=6.2.2
+#ENV GRAFANA_VERSION=5.4.3
```
Comment (NefixEstrada) [inline dockers/grafana/Dockerfile]: This comment could be deleted

File: dockers/squid/squid.conf [hunk 19]
```diff
@@ -0,0 +1,8 @@
+acl SPICE_HOSTS dst demo.isardvdi.com
```
Comment (NefixEstrada) [inline dockers/squid/squid.conf]: Should the `dst` be hardcoded rather than loaded from the environment?

File: extras/mosquitto/Dockerfile [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+VOLUME ["/mosquitto/data", "/mosquitto/log"]
+
+RUN apk add supervisor
+RUN mkdir -p /var/log/supervisor
+COPY extras/mosquitto/supervisord.conf /etc/supervisord.conf
+
+COPY extras/mosquitto/mqtt-carbon-proxy.py /
+
+CMD ["/usr/bin/supervisord", "-c", "/etc/supervisord.conf"]
+
```
Comment (NefixEstrada) [inline extras/mosquitto/Dockerfile]: Why copying the Dockerfile from someone else rather than using it as base?

File: extras/mosquitto/mqtt-carbon-proxy.py [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+            self.plugs_online[topic[3]]={}
+        self.plugs_online[topic[3]]['last_seen']=t
+        self.plugs_online[topic[3]][topic[4]]=msg.payload.decode()
+        for k,v in self.plugs_online.items():
+            if int(t-v['last_seen']) > 25:
+                # ~ del self.plugs_online[k]
+                print(k+' offline')
+        pprint.pprint(self.plugs_online)
+
+#database = db()
```
Comment (NefixEstrada) [inline extras/mosquitto/mqtt-carbon-proxy.py]: This comment could be deleted

File: extras/mosquitto/mqtt-carbon-proxy.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+database = {}
+g = grafana()
+# ~ g.send_kv('isard.hypers.isard-hypervisor.plug.power',34)
```
Comment (NefixEstrada) [inline extras/mosquitto/mqtt-carbon-proxy.py]: This comments could be deleted

File: extras/mosquitto/sample-client.py [hunk 23]
```diff
@@ -0,0 +1,15 @@
+## sudo pip3 install paho-mqtt
```
Comment (NefixEstrada) [inline extras/mosquitto/sample-client.py]: This comment could be deleted

File: src/engine/services/threads/grafana_thread.py [hunk 24]
```diff
@@ -15,27 +15,125 @@
...
+                    print('Thread running (TID: {}): {}'.format(t.tid, t.name))
+                else:
+                    print('Thread running: {}'.format(t.name))
+            return d_to_grafana
+
+        except AttributeError:
+            d_to_grafana['is_alive'] = False
+            print('ERROR ----- ENGINE IS DEATH')
+            return d_to_grafana
+        # except Exception as e:
```
Comment (NefixEstrada) [inline src/engine/services/threads/grafana_thread.py]: This comments could be deleted

File: src/engine/services/threads/grafana_thread.py [hunk 25]
```diff
@@ -109,17 +208,24 @@ def run(self):
...
                                 stats_domains_now = self.t_status[id_hyp].status_obj.hyp_obj.stats_domains_now
                             # ~ for id_domain,d_stats in stats_domains_now.items():
                             # ~ if len(stats_hyp_now) > 0:
                                 # ~ for id_domain,d_stats in stats_domains_now.items():
                                     # ~ dict_to_send[f'domain-stats-{j}'] = {'domain-id':{id_domain:1},'last': d_stats,}
                                 dict_to_send[f'hypers.'+id_hyp]['domains']=stats_domains_now #{x:0 for x in stats_domains_now}
+                                # d_radar = {}
+                                # for k in "net_tx,net_rx,vcpu_load,disk_wr,disk_rd".split(','):
+                                #     d_radar[k] = stats_domains_now[k]
+                                # dict_to_send[f'hypers.'+id_hyp]['domains_radar'] = d_radar
```
Comment (NefixEstrada) [inline src/engine/services/threads/grafana_thread.py]: This comments could be deleted

File: src/engine/services/threads/grafana_thread.py [hunk 26]
```diff
@@ -109,17 +208,24 @@ def run(self):
...
                                     # ~ print(stats_domains_now)
                                     # ~ j+=1
-
-                    if len(dict_to_send) > 0:
-                        self.send(dict_to_send)
+                #dict_to_send['engine'] = self.t_status
+                dict_to_send['engine'] = self.engine_info_to_grafana()
+                # from pprint import pprint
+                # pprint(dict_to_send)
```
Comment (NefixEstrada) [inline src/engine/services/threads/grafana_thread.py]: This comments could be deleted

File: src/engine/services/threads/grafana_thread.py [hunk 27]
```diff
@@ -109,17 +208,24 @@ def run(self):
...
+                                # d_radar = {}
+                                # for k in "net_tx,net_rx,vcpu_load,disk_wr,disk_rd".split(','):
+                                #     d_radar[k] = stats_domains_now[k]
+                                # dict_to_send[f'hypers.'+id_hyp]['domains_radar'] = d_radar
                                     # ~ print(stats_domains_now)
                                     # ~ j+=1
-
-                    if len(dict_to_send) > 0:
-                        self.send(dict_to_send)
+                #dict_to_send['engine'] = self.t_status
```
Comment (NefixEstrada) [inline src/engine/services/threads/grafana_thread.py]: This comment could be deleted

File: standalone-hypervisor.yml [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+    privileged: true
+    restart: "no"
+#networks:
+#  internet-vms:
+#    driver: macvlan
+#    driver_opts:
+#      parent: viewers
+#    ipam:
+#      config:
+#        - subnet: 172.30.0.0/16
```
Comment (NefixEstrada) [inline standalone-hypervisor.yml]: This comments  could be deleted

File: .env.original (hunk 5 above)
Comment (jvinolas) [reply .env.original]: To keep original configuration

File: docker-compose.yml (hunk 7 above)
Comment (jvinolas) [reply docker-compose.yml]: They are needed if you want to use vlans

File: CHANGELOG.md (hunk 6 above)
Comment (jvinolas) [reply CHANGELOG.md]: ok

File: docker-compose.yml (hunk 8 above)
Comment (jvinolas) [reply docker-compose.yml]: They are needed for vlans

File: dockers/app/Dockerfile (hunk 9 above)
Comment (jvinolas) [reply dockers/app/Dockerfile]: ok

File: src/engine/services/threads/grafana_thread.py (hunk 24 above)
Comment (jvinolas) [reply src/engine/services/threads/grafana_thread.py]: ok

File: src/engine/services/threads/grafana_thread.py (hunk 25 above)
Comment (jvinolas) [reply src/engine/services/threads/grafana_thread.py]: ok

File: src/engine/services/threads/grafana_thread.py (hunk 26 above)
Comment (jvinolas) [reply src/engine/services/threads/grafana_thread.py]: ok

File: src/engine/services/threads/grafana_thread.py (hunk 27 above)
Comment (jvinolas) [reply src/engine/services/threads/grafana_thread.py]: ok

File: standalone-hypervisor.yml (hunk 28 above)
Comment (jvinolas) [reply standalone-hypervisor.yml]: They are needed for vlans

File: dockers/app/Dockerfile (hunk 10 above)
Comment (jvinolas) [reply dockers/app/Dockerfile]: ok

File: dockers/app/Dockerfile (hunk 11 above)
Comment (jvinolas) [reply dockers/app/Dockerfile]: ok

File: dockers/app/requirements.pip3 (hunk 12 above)
Comment (jvinolas) [reply dockers/app/requirements.pip3]: ok

File: dockers/app/Dockerfile_quefunciona_antics (hunk 13 above)
Comment (jvinolas) [reply dockers/app/Dockerfile_quefunciona_antics]: ok

File: dockers/app/Dockerfile_upgrade (hunk 14 above)
Comment (jvinolas) [reply dockers/app/Dockerfile_upgrade]: Tests for upgrading

File: dockers/app/requirements.pip3.original (hunk 15 above)
Comment (jvinolas) [reply dockers/app/requirements.pip3.original]: tests for upgrading

File: dockers/app/requirements.pip3.quefunciona_antics (hunk 16 above)
Comment (jvinolas) [reply dockers/app/requirements.pip3.quefunciona_antics]: test for upgrading

File: dockers/app/upgraded-req.pip3 (hunk 17 above)
Comment (jvinolas) [reply dockers/app/upgraded-req.pip3]: tests for upgrading

File: dockers/grafana/Dockerfile (hunk 18 above)
Comment (jvinolas) [reply dockers/grafana/Dockerfile]: ok

File: dockers/squid/squid.conf (hunk 19 above)
Comment (jvinolas) [reply dockers/squid/squid.conf]: it is parsed by script and modified

File: extras/mosquitto/Dockerfile (hunk 20 above)
Comment (jvinolas) [reply extras/mosquitto/Dockerfile]: yes, we could do it like that

File: extras/mosquitto/mqtt-carbon-proxy.py (hunk 21 above)
Comment (jvinolas) [reply extras/mosquitto/mqtt-carbon-proxy.py]: ok

File: extras/mosquitto/mqtt-carbon-proxy.py (hunk 22 above)
Comment (jvinolas) [reply extras/mosquitto/mqtt-carbon-proxy.py]: ok

File: extras/mosquitto/sample-client.py (hunk 23 above)
Comment (jvinolas) [reply extras/mosquitto/sample-client.py]: ok

File: extras/mosquitto/Dockerfile (hunk 20 above)
Comment (jvinolas) [reply extras/mosquitto/Dockerfile]: I've just opened an issue with that. I try to keep all the images from the same alpine version, this is why I used the author original Dockerfile and modified. But we could discuss that.
